### PR TITLE
TCC - Saving types across sessions

### PR DESCRIPTION
**Pull request by XVilka**
Initial implementation - there is still a lot of work need to be done,
particularly on saving more information about types, moving platform-dependent stuff into SDBs and general code improvement (tcc's code is a mess...)

### CODE REVIEW

**Comment by radare** [comment]
pls rebase to make it grin

**Comment by radare** [comment]
please update

**Comment by XVilka** [comment]
@radare will update with more patches after this weekend.

File: shlr/tcc/libtcc.h [hunk 1]
```diff
@@ -14,7 +14,7 @@ struct TCCState;
 typedef struct TCCState TCCState;
 
 /* create a new TCC compilation context */
-LIBTCCAPI TCCState *tcc_new(void);
+LIBTCCAPI TCCState *tcc_new(const char* arch, int bits, const char *os);
```
Comment (XVilka) [inline shlr/tcc/libtcc.h]: TODO for myself - Make a structure with TCC settings, including arch, bits, os, stack size, define stack size, functions depth, etc.

**Comment by radare** [comment]
👍

File: libr/parse/code.c [hunk 2]
```diff
@@ -21,10 +22,62 @@ static void appendstring(const char *msg, char **s) {
...
+					tcc_sym_push(subtype, 0, btype);
+				}
+				free (subtype);
+				ptr = next;
+			} while (next);
+			free (members);
+		}
+		tcc_sym_push(typename, typesize, btype);
+	}
+
```
Comment (radare) [inline libr/parse/code.c]: empty line

File: libr/parse/code.c [hunk 3]
```diff
@@ -21,10 +22,62 @@ static void appendstring(const char *msg, char **s) {
...
 }
 
-R_API char *r_parse_c_file(const char *path) {
+static int typeload(void *p, const char *k, const char *v) {
+	int btype = 0;
+	CType *ctype = (CType*)malloc(sizeof(CType));
+	RAnal *anal = (RAnal*)p;
+	//r_cons_printf ("tk %s=%s\n", k, v);
+	// TODO: Add unions support
+	if (!strncmp(v, "struct", 6) && strncmp(k, "struct.", 7)) {
```
Comment (radare) [inline libr/parse/code.c]: spaces befor e(

File: libr/parse/code.c [hunk 4]
```diff
@@ -21,10 +22,62 @@ static void appendstring(const char *msg, char **s) {
 	}
 }
 
-R_API char *r_parse_c_file(const char *path) {
+static int typeload(void *p, const char *k, const char *v) {
+	int btype = 0;
+	CType *ctype = (CType*)malloc(sizeof(CType));
```
Comment (radare) [inline libr/parse/code.c]: check for null ptr, maybe use R_NEW0(CType); instead?

File: libr/parse/code.c [hunk 5]
```diff
@@ -21,10 +22,62 @@ static void appendstring(const char *msg, char **s) {
...
+				char *tmp = strchr (subtype, ',');
+				if (tmp) {
+					*tmp++ = 0;
+					tmp = strchr (tmp, ',');
+					if (tmp) {
+						*tmp++ = 0;
+					}
+					char *subname = tmp;
+					// TODO: Go recurse here
+					query = sdb_fmt(-1, "struct.%s.%s.meta", subtype, subname);
```
Comment (radare) [inline libr/parse/code.c]: space before (

File: libr/core/cmd.c [hunk 6]
```diff
@@ -455,7 +455,10 @@ R_API int r_core_run_script (RCore *core, const char *file) {
 			free (out);
 		}
 	} else if (r_parse_is_c_file (file)) {
-		char *out = r_parse_c_file (file);
+		const char *aa = r_config_get (core->config, "asm.arch");
```
Comment (radare) [inline libr/core/cmd.c]: too much spaguetti to get arch/os/bits. this info is already inside core->anal. just take it from there

File: libr/core/cmd_type.c [hunk 7]
```diff
@@ -396,7 +396,10 @@ static int cmd_type(void *data, const char *input) {
 					char *out, *tmp;
 					tmp = r_core_editor (core, NULL, "");
 					if (tmp) {
-						out = r_parse_c_string (tmp);
+						const char *aa = r_config_get (core->config, "asm.arch");
```
Comment (radare) [inline libr/core/cmd_type.c]: same here

File: libr/core/cmd_type.c [hunk 8]
```diff
@@ -405,7 +408,10 @@ static int cmd_type(void *data, const char *input) {
 						free (tmp);
 					}
 				} else {
-					char *out = r_parse_c_file (filename);
+					const char *aa = r_config_get (core->config, "asm.arch");
```
Comment (radare) [inline libr/core/cmd_type.c]: and here

File: libr/core/cmd_type.c [hunk 9]
```diff
@@ -444,7 +450,10 @@ static int cmd_type(void *data, const char *input) {
 			snprintf (tmp, sizeof (tmp) - 1, "%s;", input + 2);
 			//const char *string = input + 2;
 			//r_anal_str_to_type (core->anal, string);
-			char *out = r_parse_c_string (tmp);
+			const char *aa = r_config_get (core->config, "asm.arch");
```
Comment (radare) [inline libr/core/cmd_type.c]: and there

File: libr/parse/code.c [hunk 10]
```diff
@@ -33,10 +86,12 @@ R_API char *r_parse_c_file(const char *path) {
 	return str;
 }
 
-R_API char *r_parse_c_string(const char *code) {
+R_API char *r_parse_c_string(RAnal *anal, const char *code, const char *arch, int bits, const char* os) {
 	char *str = NULL;
```
Comment (radare) [inline libr/parse/code.c]: int bits = anal->bits
...

File: libr/parse/code.c [hunk 11]
```diff
@@ -21,10 +22,65 @@ static void appendstring(const char *msg, char **s) {
 	}
 }
 
-R_API char *r_parse_c_file(const char *path) {
+static int typeload(void *p, const char *k, const char *v) {
+	if (!p) {
+		return -1;
+	}
+	int btype = 0;
+	CType *ctype = R_NEW0(CType);
```
Comment (radare) [inline libr/parse/code.c]: missing space

File: libr/parse/code.c [hunk 12]
```diff
@@ -21,10 +22,65 @@ static void appendstring(const char *msg, char **s) {
...
+	RAnal *anal = (RAnal*)p;
+	//r_cons_printf ("tk %s=%s\n", k, v);
+	// TODO: Add unions support
+	if (!strncmp (v, "struct", 6) && strncmp(k, "struct.", 7)) {
```
Comment (radare) [inline libr/parse/code.c]: missing space

File: libr/parse/code.c [hunk 13]
```diff
@@ -21,10 +22,65 @@ static void appendstring(const char *msg, char **s) {
...
+					query = sdb_fmt (-1, "struct.%s.%s.meta", subtype, subname);
+					btype = sdb_num_get (anal->sdb_types, query, 0);
+					tcc_sym_push (subtype, 0, btype);
+				}
+				free (subtype);
+				ptr = next;
+			} while (next);
+			free (members);
+		}
+		tcc_sym_push(typename, typesize, btype);
```
Comment (radare) [inline libr/parse/code.c]: missing space

File: libr/parse/code.c [hunk 14]
```diff
@@ -33,10 +89,12 @@ R_API char *r_parse_c_file(const char *path) {
 	return str;
 }
 
-R_API char *r_parse_c_string(const char *code) {
+R_API char *r_parse_c_string(RAnal *anal, const char *code) {
 	char *str = NULL;
-	TCCState *T = tcc_new ();
+	TCCState *T = tcc_new (anal->cpu, anal->bits, anal->os);
+	if (!T) return NULL;
```
Comment (radare) [inline libr/parse/code.c]: missing braces

**Comment by radare** [comment]
there's one broken test

```
[33m[  ]  13  cmd_project: [0mCustom types in project file  
[31m[XX][0m

[35mRegression: [0mrun_tests.sh
[35mCommand: [0m /home/travis/build/radare/radare2/install/bin/radare2 -e scr.color=0 -N -q -i /tmp/r2-regressions//cmd_project-OyIz64/rad   ../bins/elf/analysis/main > /tmp/r2-regressions//cmd_project-OyIz64/out 2> /tmp/r2-regressions//cmd_project-OyIz64/err
[35mFile: [0m../bins/elf/analysis/main
[35mScript: [0mtk char **=type
tk type.char **=**z
tk type.char **.size=64
Ps custom_types > /dev/null
Pc custom_types~?char **
"td struct Foo { int bar; int x, int y }"
Ps custom_types > /dev/null
Pc custom_types~?char **
Pc custom_types~?Foo
Pd custom_types > /dev/null

[35mDiff: [0m--- /tmp/r2-regressions//cmd_project-OyIz64/exp	2017-06-01 10:30:22.983119640 +0000
+++ /tmp/r2-regressions//cmd_project-OyIz64/out	2017-06-01 10:30:23.019119640 +0000
@@ -1,3 +1,3 @@
 7
 7
-5
+8
```

**Comment by XVilka** [comment]
It's a wrong test: 1) the C syntax is wrong (comma instead of semicolon, missing semicolon) 2) I added new sdb keys (meta). Test fixed here https://github.com/radare/radare2-regressions/pull/867

**Pull request merged**